linkerd/http/detect/src/metrics.rs

```diff
@@ -0,0 +1,195 @@
+            Unit::Seconds,
+            duration.clone(),
+        );
+
+        let results = Family::default();
+        reg.register("results", "Protocol detection results", results.clone());
+
+        Self { duration, results }
+    }
+
+    pub fn metrics(&self, labels: L) -> DetectMetrics {
+        let duration = (*self.duration.get_or_create(&labels)).clone();
+
+        let not_http = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::NotHttp,
+            labels: labels.clone(),
+        }))
+        .clone();
+        let http1 = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::Http1,
+            labels: labels.clone(),
+        }))
+        .clone();
+        let h2 = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::H2,
+            labels: labels.clone(),
+        }))
+        .clone();
+        let read_timeout = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::ReadTimeout,
+            labels: labels.clone(),
+        }))
+        .clone();
+        let error = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::Error,
+            labels,
+        }))
+        .clone();
+
+        DetectMetrics {
+            duration,
+            not_http,
+            http1,
+            h2,
+            read_timeout,
+            error,
+        }
+    }
+}
+
+// === impl DetectMetrics ===
+
+impl Default for DetectMetrics {
+    fn default() -> Self {
+        Self {
+            duration: Histogram::new([0.001, 0.1].into_iter()),
+            not_http: Counter::default(),
+            http1: Counter::default(),
+            h2: Counter::default(),
+            read_timeout: Counter::default(),
+            error: Counter::default(),
+        }
+    }
+}
+
+impl DetectMetrics {
+    pub(crate) fn observe(
+        &self,
+        result: &std::io::Result<super::Detection>,
+        elapsed: time::Duration,
+    ) {
+        match result {
+            Ok(super::Detection::NotHttp) => self.not_http.inc(),
+            Ok(super::Detection::Http(Variant::Http1)) => self.http1.inc(),
+            Ok(super::Detection::Http(Variant::H2)) => self.h2.inc(),
+            Ok(super::Detection::ReadTimeout(_)) => self.read_timeout.inc(),
+            Err(_) => self.error.inc(),
+        };
+        self.duration.observe(elapsed.as_secs_f64());
+    }
+}
+
+// === impl DetectLabels ===
+
+impl<L> EncodeLabelSet for DetectLabels<L>
+where
+    L: Clone + Hash + Eq + EncodeLabelSet + Debug + Send + Sync + 'static,
+{
+    fn encode(
+        &self,
+        mut enc: prometheus_client::encoding::LabelSetEncoder,
+    ) -> Result<(), std::fmt::Error> {
+        use prometheus_client::encoding::EncodeLabel;
+
+        (
+            "result",
+            match self.result {
+                DetectResult::NotHttp => "not_http",
+                DetectResult::Http1 => "http/1",
```

Comment:
to confirm that the `/` is okay, i checked the spec here: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

> Label values may contain any Unicode characters.

this wouldn't be an acceptable label _name_, but the slash is fine for the contents of a value.